File: zenscript-code-model/src/main/java/raylras/zen/model/type/MapType.java
package raylras.zen.model.type;

import raylras.zen.model.CompilationEnvironment;
import raylras.zen.model.symbol.Operator;
import raylras.zen.model.symbol.Symbol;
import raylras.zen.model.symbol.SymbolFactory;
import raylras.zen.model.symbol.SymbolProvider;

import java.util.List;

public record MapType(Type keyType, Type valueType) implements Type, SymbolProvider {

    @Override
    public String getTypeName() {
        return valueType.getTypeName() + "[" + keyType.getTypeName() + "]";
    }

    @Override
    public String getSimpleTypeName() {
        return valueType.getSimpleTypeName() + "[" + keyType.getSimpleTypeName() + "]";
    }

    @Override
    public boolean isCastableTo(Type type, CompilationEnvironment env) {
        if (type instanceof MapType that) {
            return this.keyType.isCastableTo(that.keyType, env)
                    && this.valueType.isCastableTo(that.valueType, env);
        }
        if (type instanceof ClassType that) {
            return that.getTypeName().equals("crafttweaker.data.IData");
        }
        return Type.super.isCastableTo(type, env);
    }

    @Override
    public List<Symbol> getSymbols() {
        return SymbolFactory.builtinSymbols()
                .variable("length", IntType.INSTANCE, Symbol.Modifier.IMPLICIT_VAL)
                .variable("keys", new ArrayType(keyType), Symbol.Modifier.IMPLICIT_VAL)
                .variable("keySet", new ArrayType(keyType), Symbol.Modifier.IMPLICIT_VAL)
                .variable("values", new ArrayType(valueType), Symbol.Modifier.IMPLICIT_VAL)
                .variable("valueSet", new ArrayType(valueType), Symbol.Modifier.IMPLICIT_VAL)
                .variable("entrySet", new ArrayType(new MapEntryType(keyType, valueType)), Symbol.Modifier.IMPLICIT_VAL)
                .operator(Operator.INDEX_GET, valueType, params -> params.parameter("key", keyType))
                .operator(Operator.INDEX_SET, VoidType.INSTANCE, params -> params.parameter("key", keyType).parameter("value", valueType))
                .operator(Operator.MEMBER_GET, valueType, params -> params.parameter("key", keyType))
                .operator(Operator.MEMBER_SET, VoidType.INSTANCE, params -> params.parameter("key", keyType).parameter("value", valueType))
                .operator(Operator.FOR_IN, this)
                .build();
    }

    @Override
    public String toString() {
        return getTypeName();
    }

}

File: zenscript-code-model/src/main/kotlin/raylras/zen/model/type/MapType.kt
package raylras.zen.model.type

import raylras.zen.model.CompilationEnvironment
import raylras.zen.model.symbol.Operator
import raylras.zen.model.symbol.Symbol
import raylras.zen.model.symbol.SymbolProvider
import raylras.zen.model.symbol.symbolSequence

data class MapType(val keyType: Type, val valueType: Type) : Type, SymbolProvider {
    override val typeName by lazy { "${valueType.typeName}[${keyType.typeName}]" }

    override val simpleTypeName by lazy { "${valueType.simpleTypeName}[${keyType.simpleTypeName}]" }

    override fun isCastableTo(that: Type?, env: CompilationEnvironment): Boolean {
        return when (that) {
            is MapType -> {
                (keyType.isCastableTo(that.keyType, env)
                        && valueType.isCastableTo(that.valueType, env))
            }

            is ClassType -> {
                that.typeName == "crafttweaker.data.IData"
            }

            else -> super.isCastableTo(that, env)
        }
    }

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        return symbolSequence {
            variable("length", IntType)
            variable("keys", ArrayType(keyType))
            variable("keySet", ArrayType(keyType))
            variable("values", ArrayType(valueType))
            variable("valueSet", ArrayType(valueType))
            variable("entrySet", ArrayType(MapEntryType(keyType, valueType)))
            operator(Operator.INDEX_GET, valueType) {
                parameter("key", keyType)
            }
            operator(Operator.INDEX_SET, VoidType) {
                parameter("key", keyType)
                parameter("value", valueType)
            }
            operator(Operator.MEMBER_GET, valueType) {
                parameter("key", keyType)
            }
            operator(Operator.MEMBER_SET, VoidType) {
                parameter("key", keyType)
                parameter("value", valueType)
            }
            operator(Operator.FOR_IN, this@MapType)
        }
    }

    override fun toString() = typeName
}
